### DISCUSSION TITLE
Managing sound/music from other apps with video components

### DISCUSSION THREAD

**Discussion by Deodes** [TheWidlarzGroup/react-native-video#4946] [Ideas]
### What problem are you trying to solve?

I don't want to disable audio from other apps on some `<Video/>` components, and in some cases I need to disable it. As I see in the docs, there is an "Audio Mixing" section, which tells us to use:

```
#import <AVFoundation/AVFoundation.h>  // import

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  ...
  [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryAmbient error:nil];  // allow
  ...
}
```

in `AppDelegate.m/mm` and use the `ignoreSilentSwitch` prop (to be honest, I don't understand how it should affect this).

At first it doesn't work personally for me. The second thing is that the `mixWithOther` prop can help in case I need to mix audio, but what am I supposed to do in the case when I actually need to disable sound from other apps?

### Proposed solution

A way to manage sound from other apps manually, per `<Video/>` component - i.e. the ability to disable/enable other apps' audio on specific video components rather than globally.

#### Steps To Reproduce

1. Try to use the `mixWithOther` prop or the "Audio Mixing" topic from the docs for disabling/enabling sound from other apps.

#### Expected behaviour

1. I can manage sound from other apps manually.

### Platforms

iOS

### Version line

v6

#### Environment info

React native info output:

```bash
System:
    OS: macOS 12.3.1
    CPU: (8) arm64 Apple M1
    Memory: 460.67 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 18.0.0 - /opt/homebrew/bin/node
    Yarn: 1.22.17 - /opt/homebrew/bin/yarn
    npm: 8.6.0 - /opt/homebrew/bin/npm
    Watchman: 2022.03.21.00 - /opt/homebrew/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.0.1, iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0
    Android SDK:
      API Levels: 23, 28, 29, 30, 31
      Build Tools: 29.0.2, 30.0.2, 31.0.0
      System Images: android-29 | ARM 64 v8a, android-29 | Intel x86 Atom_64, android-29 | Google APIs ARM 64 v8a, android-29 | Google APIs Intel x86 Atom, android-29 | Google Play ARM 64 v8a, android-30 | Google APIs Intel x86 Atom, android-30 | Google Play ARM 64 v8a, android-31 | Google APIs ARM 64 v8a
      Android NDK: Not Found
  IDEs:
    Android Studio: 2020.3 AI-203.7717.56.2031.7784292
    Xcode: 13.1/13A1030d - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.11 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2 
    react-native: 0.68.1 => 0.68.1 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

Library version: 5.2.0

**Comment by BasixKOR** [comment]
I'm unsure about iOS but we have an working implementation of this on our internal patches. I'll follow up a simple PR when I got some time, just commenting here for my own tracking

**Comment by BasixKOR** [comment]
Here's what we found out by far:

- Android manages the audio playback *per playback*, as opposed to iOS managing *per app*. What it means in practice is having two `<Video />` playing simultaneously with different `mixWithOther` or `ignoreSilentSwitch` props will likely result in unintended behaviour on iOS.
- Android doesn't use the `ignoreSilentSwitch` or `mixWithOther` prop; it uses `onAudioFocusChanged` and `disableFocus` props to manage them. However, pausing/playing a video will result in audio focus change because this behaviour is hard-coded in Android native code.
  - This is what we had to patch to allow JS code to access the audio focus state.

**Comment by frankusu** [comment]
Almost same here, I'm trying to
Play background music when video muted -> stop background music when video unmuted
But even with the Appdelegate change and ignoreSilentSwitch. if I have
mixWithOther={isVideoMuted ? 'mix' : 'inherit'} the inherit doesn't apply

**Comment by sentinelweb** [comment]
i believe it should be mixWithOther={isVideoMuted ? 'mix' : 'duck'} - this should lower the background music if the foreground is unmuted

**Comment by TWG-AI** [comment]
In v7 this is handled via the `mixAudioMode` property on the player:

```ts
player.mixAudioMode = 'mixWithOthers'; // let other apps keep playing
// also: 'doNotMix' | 'duckOthers' | 'auto'
```

`ignoreSilentSwitchMode` is also available for the iOS silent-switch behavior.

> I'm an AI assistant — verify the exact values and platform notes against the v7 docs.
